### PR TITLE
Add configuration ability

### PR DESCRIPTION
**Pull request by MJKirk**

### CODE REVIEW

File: python/eos/analysis_file_description.py [hunk 1]
```diff
@@ -281,3 +282,8 @@ def from_dict(cls, **kwargs):
 #  iterations (optional): list of dicts
 #  depends-on (optional): list of strings ???????
 #  ???????
+
+# configuration schema:
+# dict, where each key is a EOS task, and the corresponding value is a list of dicts each with keys:
+#   match (mandatory): dict, whose keys are arguments of task and values are regex-ish strings
```
Comment (dvandyk) [inline python/eos/analysis_file_description.py:519]: ``of the task``

File: python/eos/analysis_file.py [hunk 2]
```diff
@@ -105,6 +105,12 @@ def __init__(self, analysis_file):
         else:
             self._steps = [StepComponent.from_dict(**s) for s in input_data['steps']]
 
+        if 'configuration' not in input_data:
+            self._configuration = {}
+        else:
+            self._configuration = {c: [ConfigurationComponent.from_dict(task=c, match=l["match"], set=l["set"]) for l in configs]
+                                   for c, configs in input_data["configuration"].items()}
```
Comment (dvandyk) [inline python/eos/analysis_file.py:112]: Can we replace ``c`` with ``task_name``?

Comment (dvandyk) [reply python/eos/analysis_file.py:112]: What about
```python3
            self._configuration = {
                task: [ConfigurationComponent.from_dict(task=task, *config) for config in configs]
                                   for task, configs in input_data["configuration"].items()
            }
```

File: python/eos/analysis_file_description.py [hunk 3]
```diff
@@ -198,6 +198,232 @@ def from_dict(cls, **kwargs):
...
+    set:dict
+
+    @classmethod
+    def from_dict(cls, **kwargs):
+        _kwargs = _copy.deepcopy(kwargs)
+        task = kwargs["task"]
+
+        # Check: nothing in match is in set
+        if (overlap := kwargs["match"].keys() & kwargs["set"].keys()): # Check for any intersection
+            raise ValueError(f'Arguments {overlap} are in both match and set for task {task}')
```
Comment (dvandyk) [inline python/eos/analysis_file_description.py:362]: ``are in both match and set of the configuration of task {task}``

File: python/eos/analysis_file_description.py [hunk 4]
```diff
@@ -198,6 +198,232 @@ def from_dict(cls, **kwargs):
...
+    base_directory:str
+
+@dataclass
+class ConfigurationComponent(Deserializable):
+    task:str
+    match:dict
+    set:dict
+
+    @classmethod
+    def from_dict(cls, **kwargs):
```
Comment (dvandyk) [inline python/eos/analysis_file_description.py:356]: Not checking that ``task``, ``match``, and ``set`` are keys of ``kwargs``.

File: python/eos/analysis_file.py [hunk 5]
```diff
@@ -441,6 +441,36 @@ def validate(self):
             print(m)
         return messages
 
+    def check_configuration(self, task_name:str, task_args:dict):
```
Comment (dvandyk) [inline python/eos/analysis_file.py:444]: ``find_configurations``?

File: python/eos/analysis_file.py [hunk 6]
```diff
@@ -441,6 +441,36 @@ def validate(self):
...
+        """Check if there is a configuration that matches the task and arguments provided."""
```
Comment (dvandyk) [inline python/eos/analysis_file.py:445]: Improve docstring, add argument descriptions and return value description.

File: python/eos/analysis_file.py [hunk 7]
```diff
@@ -441,6 +441,36 @@ def validate(self):
...
+            config_matching = remove_nones(asdict(cc.match))
+            count = 0
+            for k,v in config_matching.items():
+                try:
+                    if task_args[k] == v:
+                        count += 1
+                except KeyError:
+                    pass
+            if count == len(config_matching):
+                    potential_configs.append(cc)
```
Comment (MJKirk) [inline python/eos/analysis_file.py:465]: Indent wrong?

Comment (dvandyk) [inline python/eos/analysis_file.py:465]: Fix indentation.

File: python/eos/analysis_file_description.py [hunk 8]
```diff
@@ -419,7 +419,7 @@ def from_dict(cls, **kwargs):
             kwargs["set"].update(kwargs["match"])
             _kwargs["set"] = PredictObservablesSetDescription.from_dict(**kwargs["set"])
         else:
-            raise ValueError(f'Configuration for unknown task : {type}')
+            raise ValueError(f'Configuration for unknown task : {task}')
```
Comment (dvandyk) [inline python/eos/analysis_file_description.py:427]: This needs to be a fixup for a previous commit.

**Review by dvandyk — CHANGES_REQUESTED** [review]
Very good work. Please have a look at my comments.

**Comment by dvandyk** [comment]
> Do we want to add the configurations to the HTML representation?

We discussed and @MJKirk already create a new issue for that.

File: python/eos/analysis_file.py [hunk 9]
```diff
@@ -441,6 +454,43 @@ def validate(self):
             print(m)
         return messages
 
+    def find_configuration(self, task_name:str, task_args:dict):
+        """Find configurations (if any) that match the task and arguments provided.
```
Comment (MJKirk) [inline python/eos/analysis_file.py:458]: Specify the return values in the docstring

Comment (dvandyk) [reply python/eos/analysis_file.py:458]: Resolved.

File: python/eos/analysis_file.py [hunk 10]
```diff
@@ -441,6 +454,43 @@ def validate(self):
...
+        task_args = remove_nones(task_args)
+
+        potential_configs = []
+        # First find all the configurations that match the given task arguments
+        for cc in self._configuration[task_name]:
+            config_matching = remove_nones(asdict(cc.match))
+            count = 0
+            for k,v in config_matching.items():
+                try:
+                    if task_args[k] == v:
```
Comment (MJKirk) [inline python/eos/analysis_file.py:480]: I think here we need something more fancy, since we want to match in a regex-ish way

**Comment by dvandyk** [comment]
Should we close this one in light of #950 ?

**Comment by mreboud** [comment]
@MJKirk?

**Pull request closed without merging**